File: android/app/src/main/java/com/earendil/todonotes/data/repo/HabitEngine.kt
package com.earendil.todonotes.data.repo

import com.earendil.todonotes.data.entity.CadenceType
import com.earendil.todonotes.data.entity.Habit
import java.util.Calendar
import java.util.TimeZone

/**
 * Berechnet für ein Habit den Start der aktuellen Periode (den Reset-Punkt),
 * ab dem die Logs für den Count gezählt werden.
 *
 * Reset-Logik:
 *  - DAY / NDAYS: Periode ist [interval] Tage lang, beginnt am startDate und wiederholt sich.
 *    Reset-Punkt = startDate + k*interval Tage, wobei k so gewählt, dass der Punkt <= now
 *    und der nächste Punkt > now ist.
 *  - WEEK: Periode ist [interval] Wochen, resetted am [resetWeekday]. Woche beginnt an diesem
 *    Wochentag (z.B. Montag). Reset-Punkt = der letzte resetWeekday <= now (modulo interval Wochen).
 *  - MONTH: Periode ist [interval] Monate, resetted am [resetAnchorDay]. des Monats.
 *  - YEAR: Periode ist [interval] Jahre, resetted am [resetAnchorDay]. des [resetAnchorMonth].
 */
object HabitEngine {

    /**
     * Start der aktuellen Periode (Millis), ab dem gezählt wird.
     * Liegt immer <= now.
     */
    fun currentPeriodStart(habit: Habit, now: Long): Long {
        val nowCal = Calendar.getInstance(TimeZone.getDefault()).apply { timeInMillis = now }
        val computed = when (habit.cadenceType) {
            CadenceType.DAY -> anchoredDayStart(habit, nowCal)
            CadenceType.NDAYS -> anchoredDayStart(habit, nowCal)
            CadenceType.WEEK -> weekStart(habit, nowCal)
            CadenceType.MONTH -> monthStart(habit, nowCal)
            CadenceType.YEAR -> yearStart(habit, nowCal)
        }
        // Vor dem Startdatum gibt es keine Periode → Start ist startDate.
        return if (computed < habit.startDate) habit.startDate else computed
    }

    /**
     * Start der *nächsten* Periode nach der aktuellen (die [now] enthält).
     * Wird für „Periode abschließen & neustarten" gebraucht.
     */
    fun nextPeriodStart(habit: Habit, now: Long): Long {
        val current = currentPeriodStart(habit, now)
        val cal = Calendar.getInstance(TimeZone.getDefault()).apply { timeInMillis = current }
        when (habit.cadenceType) {
            CadenceType.DAY -> cal.add(Calendar.DAY_OF_MONTH, habit.interval)
            CadenceType.NDAYS -> cal.add(Calendar.DAY_OF_MONTH, habit.interval)
            CadenceType.WEEK -> cal.add(Calendar.WEEK_OF_YEAR, habit.interval)
            CadenceType.MONTH -> cal.add(Calendar.MONTH, habit.interval)
            CadenceType.YEAR -> cal.add(Calendar.YEAR, habit.interval)
        }
        return cal.timeInMillis
    }

    /**
     * Anzahl Logs in der aktuellen Periode. Liefert (count, periodStart, goal).
     */
    fun progress(habit: Habit, now: Long, countSinceProvider: (String, Long) -> Int): HabitProgress {
        val start = currentPeriodStart(habit, now)
        val count = countSinceProvider(habit.id, start)
        return HabitProgress(
            count = count,
            goal = habit.goalCount,
            periodStart = start
        )
    }

    // ---- DAY / NDAYS: alle [interval] Tage ab startDate ----
    private fun anchoredDayStart(habit: Habit, nowCal: Calendar): Long {
        val startCal = Calendar.getInstance(TimeZone.getDefault()).apply {
            timeInMillis = habit.startDate
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        val startMillis = startCal.timeInMillis
        if (nowCal.timeInMillis <= startMillis) return startMillis
        val dayMs = 24L * 60 * 60 * 1000
        val elapsedDays = (nowCal.timeInMillis - startMillis) / dayMs
        val periodsPassed = elapsedDays / habit.interval
        val periodStartDays = periodsPassed * habit.interval
        return startMillis + periodStartDays * dayMs
    }

    // ---- WEEK: reset am resetWeekday, alle [interval] Wochen ----
    private fun weekStart(habit: Habit, nowCal: Calendar): Long {
        val resetWd = habit.resetWeekday ?: Calendar.MONDAY
        // Zurück zum letzten resetWd <= now
        val cal = (nowCal.clone() as Calendar).apply {
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        // Differenz in Tagen bis zum letzten resetWd
        var diff = (cal.get(Calendar.DAY_OF_WEEK) - resetWd + 7) % 7
        cal.add(Calendar.DAY_OF_MONTH, -diff)
        // Jetzt haben wir den letzten resetWd (0:00). Bei interval>1 müssen wir noch
        // ganze Wochen zurückspringen, damit wir in der "richtigen" Woche landen.
        if (habit.interval > 1) {
            val startCal = Calendar.getInstance(TimeZone.getDefault()).apply {
                timeInMillis = habit.startDate
                set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
                // auf den resetWd der Startwoche normieren
                val d0 = (get(Calendar.DAY_OF_WEEK) - resetWd + 7) % 7
                add(Calendar.DAY_OF_MONTH, -d0)
            }
            val weeksBetween = weeksBetween(startCal, cal)
            val periodsPassed = weeksBetween / habit.interval
            val periodStartWeeks = periodsPassed * habit.interval
            cal.timeInMillis = startCal.timeInMillis
            cal.add(Calendar.WEEK_OF_YEAR, periodStartWeeks.toInt())
        }
        return cal.timeInMillis
    }

    private fun weeksBetween(a: Calendar, b: Calendar): Long {
        val aMs = a.timeInMillis
        val bMs = b.timeInMillis
        val dayMs = 24L * 60 * 60 * 1000
        return ((bMs - aMs) / dayMs) / 7
    }

    // ---- MONTH: reset am resetAnchorDay., alle [interval] Monate ----
    private fun monthStart(habit: Habit, nowCal: Calendar): Long {
        val anchorDay = habit.resetAnchorDay ?: 1
        val cal = (nowCal.clone() as Calendar).apply {
            set(Calendar.DAY_OF_MONTH, anchorDay)
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        if (cal.timeInMillis > nowCal.timeInMillis) {
            cal.add(Calendar.MONTH, -1)
        }
        // Bei interval>1: zurückspringen in die richtige Periode relativ zum startDate
        if (habit.interval > 1) {
            val startCal = Calendar.getInstance(TimeZone.getDefault()).apply {
                timeInMillis = habit.startDate
                set(Calendar.DAY_OF_MONTH, anchorDay)
                set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
                if (timeInMillis > habit.startDate) add(Calendar.MONTH, -1)
            }
            val monthsBetween = monthsBetween(startCal, cal)
            val periodsPassed = monthsBetween / habit.interval
            val periodStartMonths = periodsPassed * habit.interval
            cal.timeInMillis = startCal.timeInMillis
            cal.add(Calendar.MONTH, periodStartMonths.toInt())
        }
        return cal.timeInMillis
    }

    private fun monthsBetween(a: Calendar, b: Calendar): Long {
        val yearDiff = b.get(Calendar.YEAR) - a.get(Calendar.YEAR)
        val monthDiff = b.get(Calendar.MONTH) - a.get(Calendar.MONTH)
        return yearDiff * 12L + monthDiff
    }

    // ---- YEAR: reset am resetAnchorDay. des resetAnchorMonth, alle [interval] Jahre ----
    private fun yearStart(habit: Habit, nowCal: Calendar): Long {
        val anchorDay = habit.resetAnchorDay ?: 1
        val anchorMonth = (habit.resetAnchorMonth ?: 0) - 1 // 0-based für Calendar
        val cal = (nowCal.clone() as Calendar).apply {
            set(Calendar.MONTH, anchorMonth)
            set(Calendar.DAY_OF_MONTH, anchorDay)
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        if (cal.timeInMillis > nowCal.timeInMillis) {
            cal.add(Calendar.YEAR, -1)
        }
        if (habit.interval > 1) {
            val startCal = Calendar.getInstance(TimeZone.getDefault()).apply {
                timeInMillis = habit.startDate
                set(Calendar.MONTH, anchorMonth)
                set(Calendar.DAY_OF_MONTH, anchorDay)
                set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
                if (timeInMillis > habit.startDate) add(Calendar.YEAR, -1)
            }
            val yearsBetween = (cal.get(Calendar.YEAR) - startCal.get(Calendar.YEAR)).toLong()
            val periodsPassed = yearsBetween / habit.interval
            val periodStartYears = periodsPassed * habit.interval
            cal.timeInMillis = startCal.timeInMillis
            cal.add(Calendar.YEAR, periodStartYears.toInt())
        }
        return cal.timeInMillis
    }
}

data class HabitProgress(
    val count: Int,
    val goal: Int,
    val periodStart: Long
) {
    val isComplete: Boolean get() = count >= goal
    val ratio: Float get() = if (goal <= 0) 0f else count.toFloat() / goal
}

File: archive/linux-gtk4/todonotes/habit_engine.py
"""Habit Engine — Port des Kotlin HabitEngine.

Berechnet den Start der aktuellen Periode (Reset-Punkt) für ein Habit.
Reset-Logik identisch zu Android:
  - DAY / NDAYS: Periode ist [interval] Tage, beginnt am startDate.
  - WEEK: reset am resetWeekday, alle [interval] Wochen.
  - MONTH: reset am resetAnchorDay, alle [interval] Monate.
  - YEAR: reset am resetAnchorDay des resetAnchorMonth, alle [interval] Jahre.

Python datetime statt java.util.Calendar.
Wochentag-Mapping: Android Calendar.MONDAY=2 ... SUNDAY=1.
Python: Monday=0 ... Sunday=6.
"""
from __future__ import annotations

from datetime import datetime, timedelta

# Android Calendar weekday → Python weekday (0=Monday)
# Calendar.SUNDAY=1, MONDAY=2, TUESDAY=3, WEDNESDAY=4, THURSDAY=5, FRIDAY=6, SATURDAY=7
_ANDROID_TO_PY_WEEKDAY = {
    1: 6,  # Sunday → 6
    2: 0,  # Monday → 0
    3: 1,  # Tuesday → 1
    4: 2,  # Wednesday → 2
    5: 3,  # Thursday → 3
    6: 4,  # Friday → 4
    7: 5,  # Saturday → 5
}


def _to_date(millis: int) -> datetime:
    return datetime.fromtimestamp(millis / 1000)


def _to_millis(dt: datetime) -> int:
    return int(dt.timestamp() * 1000)


def _midnight(dt: datetime) -> datetime:
    return dt.replace(hour=0, minute=0, second=0, microsecond=0)


def current_period_start(habit: dict, now: int) -> int:
    """Start der aktuellen Periode (Millis), ab dem gezählt wird. Liegt <= now."""
    cadence = habit["cadenceType"]
    now_dt = _to_date(now)

    if cadence in ("DAY", "NDAYS"):
        computed = _anchored_day_start(habit, now_dt)
    elif cadence == "WEEK":
        computed = _week_start(habit, now_dt)
    elif cadence == "MONTH":
        computed = _month_start(habit, now_dt)
    elif cadence == "YEAR":
        computed = _year_start(habit, now_dt)
    else:
        computed = _to_date(habit["startDate"])

    # Vor dem Startdatum → Start ist startDate
    start_millis = habit["startDate"]
    result = _to_millis(computed)
    return result if result >= start_millis else start_millis


def next_period_start(habit: dict, now: int) -> int:
    """Start der nächsten Periode nach der aktuellen."""
    current = _to_date(current_period_start(habit, now))
    interval = habit["interval"]
    cadence = habit["cadenceType"]

    if cadence in ("DAY", "NDAYS"):
        nxt = current + timedelta(days=interval)
    elif cadence == "WEEK":
        nxt = current + timedelta(weeks=interval)
    elif cadence == "MONTH":
        nxt = _add_months(current, interval)
    elif cadence == "YEAR":
        nxt = _add_months(current, interval * 12)
    else:
        nxt = current + timedelta(days=interval)

    return _to_millis(nxt)


def progress(
    habit: dict, now: int, count_since_fn
) -> tuple[int, int, int]:
    """(count, goal, period_start)."""
    start = current_period_start(habit, now)
    count = count_since_fn(habit["id"], start)
    return (count, habit["goalCount"], start)


# ── DAY / NDAYS ─────────────────────────────────────────────────

def _anchored_day_start(habit: dict, now_dt: datetime) -> datetime:
    start_dt = _midnight(_to_date(habit["startDate"]))
    if now_dt <= start_dt:
        return start_dt
    day = timedelta(days=1)
    elapsed_days = (now_dt - start_dt).days
    periods_passed = elapsed_days // habit["interval"]
    period_start_days = periods_passed * habit["interval"]
    return start_dt + timedelta(days=period_start_days)


# ── WEEK ───────────────────────────────────────────────────────

def _week_start(habit: dict, now_dt: datetime) -> datetime:
    reset_wd_android = habit.get("resetWeekday") or 2  # Monday default
    reset_wd = _ANDROID_TO_PY_WEEKDAY[reset_wd_android]

    cal = _midnight(now_dt)
    # Zurück zum letzten resetWd
    diff = (cal.weekday() - reset_wd) % 7
    cal = cal - timedelta(days=diff)

    if habit["interval"] > 1:
        start_cal = _midnight(_to_date(habit["startDate"]))
        d0 = (start_cal.weekday() - reset_wd) % 7
        start_cal = start_cal - timedelta(days=d0)
        weeks_between = (cal - start_cal).days // 7
        periods_passed = weeks_between // habit["interval"]
        period_start_weeks = periods_passed * habit["interval"]
        cal = start_cal + timedelta(weeks=period_start_weeks)

    return cal


# ── MONTH ──────────────────────────────────────────────────────

def _month_start(habit: dict, now_dt: datetime) -> datetime:
    anchor_day = habit.get("resetAnchorDay") or 1
    cal = now_dt.replace(day=anchor_day)
    cal = _midnight(cal)
    if cal > now_dt:
        cal = _add_months(cal, -1)

    if habit["interval"] > 1:
        start_cal = _to_date(habit["startDate"]).replace(day=anchor_day)
        start_cal = _midnight(start_cal)
        if start_cal.timestamp() * 1000 > habit["startDate"]:
            start_cal = _add_months(start_cal, -1)
        months_between = _months_between(start_cal, cal)
        periods_passed = months_between // habit["interval"]
        period_start_months = periods_passed * habit["interval"]
        cal = _add_months(start_cal, period_start_months)

    return cal


# ── YEAR ───────────────────────────────────────────────────────

def _year_start(habit: dict, now_dt: datetime) -> datetime:
    anchor_day = habit.get("resetAnchorDay") or 1
    anchor_month = (habit.get("resetAnchorMonth") or 1) - 1  # 0-based
    cal = now_dt.replace(month=anchor_month + 1, day=anchor_day)
    cal = _midnight(cal)
    if cal > now_dt:
        cal = _add_months(cal, -12)

    if habit["interval"] > 1:
        start_cal = _to_date(habit["startDate"]).replace(
            month=anchor_month + 1, day=anchor_day
        )
        start_cal = _midnight(start_cal)
        if start_cal.timestamp() * 1000 > habit["startDate"]:
            start_cal = _add_months(start_cal, -12)
        years_between = cal.year - start_cal.year
        periods_passed = years_between // habit["interval"]
        period_start_years = periods_passed * habit["interval"]
        cal = _add_months(start_cal, period_start_years * 12)

    return cal


# ── Helpers ─────────────────────────────────────────────────────

def _add_months(dt: datetime, months: int) -> datetime:
    """Addiert Monate (kann Overflow am Monatsende geben → Tag anpassen)."""
    total = dt.month - 1 + months
    year = dt.year + total // 12
    month = total % 12 + 1
    # Tag ggf. kürzen (z.B. 31. Jan + 1 Monat → 28. Feb)
    import calendar
    max_day = calendar.monthrange(year, month)[1]
    day = min(dt.day, max_day)
    return dt.replace(year=year, month=month, day=day)


def _months_between(a: datetime, b: datetime) -> int:
    return (b.year - a.year) * 12 + (b.month - a.month)


def cadence_label(habit: dict) -> str:
    """Kurzes Label wie '2x pro Woche' / '1x alle 3 Tage'."""
    cadence = habit["cadenceType"]
    per = {
        "DAY": "Tag",
        "WEEK": "Woche",
        "MONTH": "Monat",
        "YEAR": "Jahr",
        "NDAYS": f"{habit['interval']} Tage",
    }.get(cadence, cadence)
    if cadence == "NDAYS":
        return f"{habit['goalCount']}x alle {per}"
    return f"{habit['goalCount']}x pro {per}"
